### DISCUSSION TITLE
Question about coreclr, hostfxr and nethost in net 8 or 9

### DISCUSSION THREAD

**Discussion by DeafMan1983** [dotnet/runtime#110885] [Q&A] [answered]
Hello everyone, I want to know.

But I don't understand why does @tannergooding write old example from old repository about native-hostfxr but I really want to know how do I load assembly inside dotnet 8 or 9 because I want create simple embedded library ( libdotnetapp_interop.a ) in Dotnet's Console

1. Create function LoadAssembly(const char* app_path, const char* type_name, const char* method_name, void** delegate_pointer)
2. Compile with gcc and ar into `*.a`!
3. Copy *.a to DotnetApp like pure C# application!
4. DirectInvoke Include="__Internal"
   NativeLibrary Include="-L./ -ldotnetapp_interop -lnethost -lcoreclr -lhostfxr" 
5. Write `DllImport`:
```
[DllImport("__Internal")]
public unsafe static extern void LoadAndExecuteAssembly(sbyte* assembly_path, sbyte* type_name, sbyte* method_name, void** dele_pointers);
```

But I really want to know how do I get success initializing if I get "Failed to initialize coreclr!" But I don't know. But I never make statically with libraries of Net/Dotnet SDK.

I really want to have own example if I understand correctly.

Make sure example of **implemented function from C/C++**!
And managed assembly should execute by implemented function.

What do I need know for app_domain?

Example Managed assembly:
```
namespace MyAssembly;

public class HelloTest
{
    public static void Hello()
    {
        Console.WriteLine("Hello from MyAssembly!");
    }

    public static string ShowString()
    {
        return "That is a managed assembly.";
    }

    /*
        Any implemented functions like managed assembly loads or whatever
    */
}
```

But Unmanaged P/invokes I know that. like in C/C++:
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <nethost.h>
#include <hostfxr.h>

extern void SayPrint(void) {
    printf("From unmanaged print\n");
}

// Implement LoadAndExecuteAssembly()
...
```

Because I want to learn about it. Thank you!

**Comment by huoyaoyuan** [comment]
It's unclear about what do you want to achieve. To better express your intend and failures, you can highlight the following points:

How many module of code are you trying to connect together, 2 or 3? What's the language and runtime environment of each module? What's the interface you were using to load them?

**Reply by DeafMan1983** [reply]
I mean I want to load assembly from coreclr ( embedded inside libdotnetapp_interop.a, but it should be dynamical with 2 shared libraries libcoreclr.so and libhostfxr.so )

And DotnetApp should load assembly ( MyAssembly.dll only C# as managed assembly )

I am using for C ( not C++ ) for static library. And managed assembly uses only C#.

DotnetApp is dotnet console project ( pure C# ) with achieve libdotnetapp_interop.a for __Interal. It works like Godot or Unity3D.

Sorry my bad English! Because I am a German. :/
Just my working project for example:
![image](https://github.com/user-attachments/assets/2b2482f0-9d71-4e16-aaff-e29e8ab1ea14)
I am trying to create a communication with bridge managed and native implementations. That is all. But it is my inofficial development. Because I expect somebody said Blender want to move to C# but they won't release. I try to port to C# version. I missed to communicate with "bridges" with managed and native functions ( like static methods, UnmanagedCallerOnly or DllImport ).

**Reply by huoyaoyuan** [reply, marked as answer]
To summarize:

Your host and entry application is native and developed in C. The CLR instance is hosted by the C application. The bridge should be established via either nethost.h, or nativeaot.

The error message "Failed to initialize coreclr!" sounds like there's another CLR instance already running in the process, used by any service. CLR doesn't support co-existence within the same process. In this case, using nativeaot should be the only option.

**Comment by tannergooding** [comment]
> But I don't understand why does @tannergooding write old example from old repository about native-hostfxr but I really want to know how do I load assembly inside dotnet 8 or 9 because I want create simple embedded library ( libdotnetapp_interop.a ) in Dotnet's Console

I'm unsure what sample you're referring to, can you provide a link?

**Reply by DeafMan1983** [reply]
I mean I found hostfxr sample under dotnet github repository but I don't understand why does coreclr not work for net 8.0+? But it always say "Failed to initialize coreclr" if I use Linux Mint. libcoreclr.so found in dotnet 8.0 SDK /user/lib/dotnet/Microsoft.Net.Core.App/runtime or where path to libcoreclr.so.

How do I embed all assemblies into nativeapp? It works like mkbundle of mono runtime? Similar to use `dotnet publish -c Release -r linux-x64 --self-contained=true -p:PublishAot=true`?
